### PR TITLE
Allow time tracking entries to be edited

### PR DESCRIPTION
**Pull request by balsdorf**
Either the user who submitted them or a manager can edit.

### CODE REVIEW

File: lib/eventum/class.time_tracking.php [hunk 1]
```diff
@@ -396,6 +396,41 @@ function ($a, $b) {
...
+                 WHERE
+                    ttr_ttc_id=ttc_id AND
+                    ttr_usr_id=usr_id AND
+                    ttr_id=?';
+        try {
+            $res = DB_Helper::getInstance()->getRow($stmt, [$ttr_id]);
+        } catch (DatabaseException $e) {
+            return false;
+        }
+
```
Comment (glensc) [inline lib/eventum/class.time_tracking.php]: new code should just let exceptions to be thrown and handle them as needed

File: lib/eventum/class.time_tracking.php [hunk 2]
```diff
@@ -518,6 +553,64 @@ public static function addTimeEntry($iss_id, $ttc_id, $time_spent, $date, $summa
...
+            $time_spent,
+            $summary,
+            $ttr_id
+        ];
+        try {
+            DB_Helper::getInstance()->query($stmt, $params);
+        } catch (DatabaseException $e) {
+            return -1;
+        }
+
```
Comment (glensc) [inline lib/eventum/class.time_tracking.php]: same here about exceptions

File: src/Controller/TimeTrackingController.php [hunk 3]
```diff
@@ -99,5 +135,14 @@ protected function prepareTemplate()
                 'time_categories' => Time_Tracking::getAssocCategories($prj_id),
             ]
         );
+
+        if ($this->time_tracking_details) {
+            $this->tpl->assign([
+                'details'   =>  $this->time_tracking_details,
+                'start_date'    =>  \Date_Helper::getDateTime($this->time_tracking_details['ttr_created_date']),
+                'end_date'    =>  \Date_Helper::getDateTime($this->time_tracking_details['ttr_created_date'])->sub(
```
Comment (glensc) [inline src/Controller/TimeTrackingController.php]: `use Date_Helper` i prefer if you don't fully qualify the class names

File: templates/time_tracking_entry.tpl.html [hunk 4]
```diff
@@ -79,58 +88,70 @@
...
+        <input type="hidden" name="ttr_id" value="{$details.ttr_id}">
+        {else}
         <input type="hidden" name="cat" value="add_time">
         <input type="hidden" name="issue_id" value="{$issue_id}">
+        {/if}
         <table class="bordered">
             {if $core.current_role > $core.roles.viewer}
             <tr class="title">
-                <th colspan="2">{t}Record Time Worked{/t}</th>
+                <th colspan="2">{t 1=$action}%1 Time Worked{/t}</th>
```
Comment (glensc) [inline templates/time_tracking_entry.tpl.html]: this is not translatable, make two different `{t}` placeholders. assuming that word order in sentence is same in all languages is naïve

File: templates/time_tracking_entry.tpl.html [hunk 5]
```diff
@@ -79,58 +88,70 @@
...
+                        time=$start_date|default:''}&nbsp;
+                    {html_select_time minute_interval=5 field_array='date2' prefix='' all_extra=' class="default" onChange="Eventum.calcDateDiff(this.form, 2, this.name)"' display_seconds=false
+                        time=$end_date|default:''}
+                    <a href="javascript:void(null);" onClick="Eventum.updateTimeFields($('form#add_time_form'), 'date2[Year]', 'date2[Month]', 'date2[Day]', 'date2[Hour]', 'date2[Minute]'); Eventum.calcDateDiff($('form#add_time_form'), 12)"><img src="images/icons/refresh.gif" border="0"></a>
                 </td>
             </tr>
             <tr class="buttons">
                 <td colspan="2">
-                    <input type="submit" value="{t}Add Time Entry{/t}">
+                    <input type="submit" value="{t 1=$action}%1 Time Entry{/t}">
```
Comment (glensc) [inline templates/time_tracking_entry.tpl.html]: same here

File: src/Controller/TimeTrackingController.php (hunk 3 above)
Comment (balsdorf) [inline src/Controller/TimeTrackingController.php]: IDE auto did that, fixed.

File: templates/time_tracking_entry.tpl.html (hunk 4 above)
Comment (balsdorf) [inline templates/time_tracking_entry.tpl.html]: Ugh, I can't believe I did that. Thanks. Fixed

**Comment by slay123** [comment]
problem no 1.
cannot delete timetrackings

**Comment by slay123** [comment]
problem no 2.
when closing issue with timetracking, issue will be closed 2 times and 2 time tracking entries are being inserted

**Comment by balsdorf** [comment]
@slay123 I've fixed the bug with deleting the time entry. However I cannot reproduce your problem with closing issues. Do you have a workflow enabled that could be causing this?

**Comment by slay123** [comment]
I couldn't re-produce problem 2. perhaps my Firefox was lagging and caused the problem.

**Pull request merged**